Add gcom api processor.

kovrus opened the pull request:
See [Design Doc: Use OTel Collector as a replacement for OTLP Gateway in Grafana Cloud OTLP Endpoint
](https://docs.google.com/document/d/1HsJr5eVH4WOdSSGIeaYRRUGAqx4Bzx-CSna1mspW4a4/edit#heading=h.5sybau7waq2q) for more information on why this component is needed.

Note: I've copied only cache and client modules from `https://github.com/grafana/backend-enterprise`. I did not use it as dep because it causes some des conflicts and it has to be fetched from private repo (which might be ok actually).   

Follow up:
- add `otlp-gateway` distribution with this and the routing processor component. ( https://github.com/grafana/opentelemetry-collector-components/pull/46/commits/fa135124bff990305a487e7692319ab3704f0ad1)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/opentelemetry-collector-components?pullRequest=44) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/opentelemetry-collector-components?pullRequest=44) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/opentelemetry-collector-components?pullRequest=44) it.</sub>

kovrus commented:
`internal.{cache, client, common}` should not be reviewed, it is one to one copy from  `https://github.com/grafana/backend-enterprise`

jpkrohling commented on components/processor/gcomapiprocessor/internal/gcom/cache/auth_cache.go:
```diff
@@ -0,0 +1,96 @@
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/opentelemetry-collector-components/processor/gcomapiprocessor/internal/gcom/client"
```
we might need a linter to enforce three import groups -- this one here would belong to a third group, of local imports

jpkrohling commented on components/processor/gcomapiprocessor/internal/gcom/cache/auth_cache.go:
```diff
@@ -0,0 +1,96 @@
+
+// CheckKey checks if the provided key is valid
+func (a *AuthCache) CheckKey(ctx context.Context, key string) (*client.APIKey, error) {
+	span, _ := opentracing.StartSpanFromContext(ctx, "Gateway.CheckKey")
```
I know which code I'll change from opentracing to otel first :-)

jpkrohling commented on components/processor/gcomapiprocessor/internal/gcom/cache/auth_cache.go:
```diff
@@ -0,0 +1,96 @@
+
+type AuthCache struct {
+	client   client.Client
+	logger   log.Logger
```
The collerctor uses zap instead -- does it make sense to switch to it at a follow-up PR?

jpkrohling commented on components/processor/gcomapiprocessor/internal/gcom/cache/instance_cache.go:
```diff
@@ -0,0 +1,290 @@
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet
+func (cfg *InstanceCacheConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
```
Are those flags actually used?

jpkrohling commented on components/processor/gcomapiprocessor/internal/gcom/cache/instance_cache.go:
```diff
@@ -0,0 +1,290 @@
+
+	if len(instanceTypes) == 0 {
+		// TODO (@chroberts): replace this with an error after we roll out
+		// https://github.com/grafana/backend-enterprise/pull/3749
```
This is merged -- should this code be changed already?

jpkrohling commented on components/processor/gcomapiprocessor/internal/gcom/cache/instance_cache.go:
```diff
@@ -0,0 +1,290 @@
+}
+
+func (i *instanceCache) run() {
+	completeRefreshTicker := time.NewTicker(i.cfg.CompleteCacheRefreshDuration)
```
Ideally, this would receive a context, which would be cancelled when the component is shutting down. In that case, the for range channel would be replaced with selects, so that we can cancel the cache refresh operations when we are shutting down.

jpkrohling commented on components/processor/gcomapiprocessor/internal/gcom/client/api_key.go:
```diff
@@ -0,0 +1,110 @@
+
+	res, err := c.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("err=%w, msg=%v", ErrAccess, err)
```
Are we leaking internal details with this message? Or do we trust that the internal gateway is returning only information that is safe for the general public?

jpkrohling commented on components/processor/gcomapiprocessor/internal/gcom/client/client.go:
```diff
@@ -0,0 +1,154 @@
+)
+
+var (
+	grafanaComReqs = promauto.NewHistogramVec(
```
TODO: change this to otel, not sure if those would join the ones that are generated by the collector itself. In the future, the collector might not export automatically to Prometheus like it does today.

jpkrohling replied:
Do we want to use the telemetry settings that the collector provides? While the registries might end up merged, this is likely to change in the future.

jpkrohling commented on components/processor/gcomapiprocessor/internal/gcom/client/client.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
```
Should those flags be removed?

jpkrohling replied:
Should we remove this?

jpkrohling commented on components/processor/gcomapiprocessor/internal/gcom/client/client.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+// New creates a grafana.com Client
+func New(cfg Config, name string, logger log.Logger) (Client, error) {
```
I just realized that those are the two packages you said you just copied from the enterprise package. If we are going to maintain them, should we accomodate them to the collector standards? Or do we want to threat those two packages as libraries? In that case, it would probably make sense to split them into different go modules perhaps.

kovrus replied:
I'll follow up with a PR against backend-enterprise to convert those packages to modules. If it wouldn't work for some reason, let's improve the code and strip all unnecessary parts.

jpkrohling commented on components/processor/gcomapiprocessor/testdata/config.yaml:
```diff
@@ -0,0 +1,17 @@
+    key: "test_key"
+    timeout: 1s
+  cache:
+    complete_cache_refresh_duration: 1s
```
complete_refresh_duration -- "cache" is already the parent's node name

jpkrohling commented on components/processor/gcomapiprocessor/testdata/config.yaml:
```diff
@@ -0,0 +1,17 @@
+    timeout: 1s
+  cache:
+    complete_cache_refresh_duration: 1s
+    incremental_cache_refresh_duration: 1s
```
incremental_refresh_duration

jpkrohling commented on components/processor/gcomapiprocessor/README.md:
````diff
@@ -0,0 +1,67 @@
+```
+
+[alpha]: https://github.com/open-telemetry/opentelemetry-collector#alpha
+[design doc]: https://docs.google.com/document/d/1HsJr5eVH4WOdSSGIeaYRRUGAqx4Bzx-CSna1mspW4a4/edit#heading=h.89ldx0hih690
````
This is Grafana-internal. Is this a problem? Not sure myself.

kovrus replied:
The component is Grafana internal as well, so linking this design doc should be fine I guess.

jpkrohling commented on components/processor/gcomapiprocessor/config.go:
```diff
@@ -0,0 +1,38 @@
+)
+
+type Config struct {
+	serviceName string       `mapstructure:"service_name"`
```
Shouldn't this be in upper case?

jpkrohling commented on components/processor/gcomapiprocessor/config.go:
```diff
@@ -0,0 +1,38 @@
+	serviceName string       `mapstructure:"service_name"`
+	Client      clientConfig `mapstructure:"client"`
+	Cache       cacheConfig  `mapstructure:"cache"`
+	IsTest      bool         `mapstructure:"test"`
```
Dry-run? Mock? Dev? Perhaps we don't even need this if we provide a container image with a gcom API mock?

jpkrohling replied:
Or perhaps we could have a special handling for an endpoint that starts with "mock://" ?

kovrus replied:
> perhaps we don't even need this if we provide a container image with a gcom API mock?

> Or perhaps we could have a special handling for an endpoint that starts with "mock://" ?

I like both options, but would probably lean towards `mock://` since the test setup will be simpler.

jpkrohling commented on components/processor/gcomapiprocessor/go.mod:
```diff
@@ -0,0 +1,61 @@
+	go.uber.org/zap v1.24.0
+)
+
+require (
```
Why do we have three require groups?

jpkrohling commented on components/processor/gcomapiprocessor/processor.go:
```diff
@@ -0,0 +1,116 @@
+	if err != nil {
+		return nil, err
+	}
+	if cfg.IsTest {
```
Here we could check for the "mock://" ?

jpkrohling commented on components/processor/gcomapiprocessor/processor.go:
```diff
@@ -0,0 +1,116 @@
+	v := info.Metadata.Get(orgID)
+
+	if v == nil || len(v) == 0 {
+		return nil, fmt.Errorf("missing '%s' header", orgID)
```
You could have used %q here: "missing %q header"

jpkrohling commented on go.sum:
```diff
@@ -2,8 +2,6 @@ github.com/caarlos0/log v0.1.2 h1:P/pppG5Pv2wfapmioBCQaydFUteMbOy9nSttUHbdowY=
 github.com/caarlos0/testfs v0.4.4 h1:3PHvzHi5Lt+g332CiShwS8ogTgS3HjrmzZxCm6JCDr8=
 github.com/caarlos0/testfs v0.4.4/go.mod h1:bRN55zgG4XCUVVHZCeU+/Tz1Q6AxEJOEJTliBy+1DMk=
-github.com/charmbracelet/lipgloss v0.5.1-0.20220615005615-2e17a8a06096 h1:ai19sA3Zyg3DARevWCbdLOWt+MfWiE3e8voBqzFOgP8=
-github.com/charmbracelet/lipgloss v0.5.1-0.20220615005615-2e17a8a06096/go.mod h1:D7uPgcyfB9T1Ug2mfJOnES17o47nz5oqIzSSVrpcviU=
```
This does not seem related?

jpkrohling reviewed (approved):
I'm OK with merging this as is, as I believe you tested and said this worked already. Everything I listed here can be done later on follow-up PRs.

kovrus commented:
I've addressed all the comments except comments on the `internal` package, I'd like to try to create client and cache modules in backend enterprise first, if it doesn't work, I'll address those comments too in a separate PR. For now, I'll proceed with merging this PR.
